API: Pass TokenMetadataError through to API

rvl opened the pull request:
### Issue Number

ADP-925

### Overview

If there was an error fetching metadata in the listMetadata or getMetadata endpoints, tell the user, rather than just returning "no metadata".

### Comments

Could someone please pick this up and merge for v2021-05-26.

rvl committed:
API: Pass TokenMetadataError through to API

rvl committed:
integration tests: Add metadata_error field checks

Anviking commented on lib/core/src/Cardano/Wallet/TokenMetadata.hs:
```diff
@@ -356,6 +355,9 @@ data TokenMetadataError
     deriving (Generic, Show, Eq)
 
+instance NFData TokenMetadataError where
+    rnf = rnf . show
```
Is this `show` right? 🤔

rvl replied:
Yes it's a cheap trick to make `NFData` instances, where some of the fields don't have `NFData`.
In this case the problem was `SomeException`.
`show` usually works well enough for the purpose of deeply evaluating some data.

Anviking committed:
Re-generate ApiAsset goldens with basic metadata

This commit adds very basic generators for token metadata (better than
the not generating at all), and re-regnerates the golden.

Now we see the new metadata_error field in some goldens.

Anviking committed:
Add ApiMetadataError type

Anviking commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -480,7 +480,8 @@ data ApiAsset = ApiAsset
     , fingerprint :: ApiT W.TokenFingerprint
     , metadata :: Maybe ApiAssetMetadata
-    } deriving (Eq, Generic, Ord, Show)
+    , metadataError :: Maybe Text
```
Replaced with a proper sum-type

Anviking reviewed (approved):
lgtm

Testing this seems a bit cumbersome, as I think our mock metadata server would refuse to serve invalid JSON.

rvl commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1791,8 +1799,37 @@ instance Arbitrary Word31 where
+    arbitrary = AssetDecimals <$> choose (1, 19)
+
+instance Arbitrary AssetMetadata where
+    -- TODO: We should add a proper arbitrary instance
```
Thanks, yes we should.

Anviking commented on lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:
```diff
@@ -1791,8 +1799,37 @@ instance Arbitrary Word31 where
+        <*> (pure "An asset")
+        <*> (oneof [pure Nothing, pure $ Just "AST"])
+        <*> genMaybe (pure $ AssetURL $ fromJust $
+                parseURI "https://asset.url")
```
FYI - I first tried using an existing `Aribitrary URI` instance, but it failed, because it generated `http` urls despite the swagger api promising it to be `https`. 

I'm not sure what the metadata server promises, but if it allows `http` urls, then our swagger spec promises too much...
